lib/rules/no-hooks-from-ancestor-modules.js

```diff
@@ -0,0 +1,106 @@
+
+        //----------------------------------------------------------------------
+        // Helpers
+        //----------------------------------------------------------------------
+
+        function isInModuleCallbackBody(callExpressionNode) {
+            return callExpressionNode &&
+                callExpressionNode.parent &&
+                callExpressionNode.parent.type === "ExpressionStatement" &&
+                callExpressionNode.parent.parent &&
+                callExpressionNode.parent.parent.type === "BlockStatement" &&
+                callExpressionNode.parent.parent.parent &&
+                callExpressionNode.parent.parent.parent.type === "FunctionExpression" &&
+                callExpressionNode.parent.parent.parent.parent &&
+                callExpressionNode.parent.parent.parent.parent.type === "CallExpression" &&
+                utils.isModule(callExpressionNode.parent.parent.parent.parent.callee);
+        }
+
+        function isHookInvocation(node) {
+            return node.callee.type === "MemberExpression" &&
+                node.callee.object.type === "Identifier" &&
+                NESTABLE_HOOK_NAMES.includes(node.callee.property.name) &&
+                isInModuleCallbackBody(node);
+        }
+
+        function isFunctionWithIdentifierArg(node) {
+            return node &&
+                node.type === "FunctionExpression" &&
+                node.params.length > 0 &&
+                node.params[0].type === "Identifier";
+        }
+
+        //----------------------------------------------------------------------
+        // Public
+        //----------------------------------------------------------------------
+
+        return {
+            "CallExpression": function (node) {
+                if (utils.isModule(node.callee)) {
+                    contextStack.push("module");
+
+                    const callback = node.arguments[1];
+                    if (isFunctionWithIdentifierArg(callback)) {
+                        hooksStack.push(callback.params[0].name);
+                    } else {
+                        hooksStack.push(null);
+                    }
+                } else if (isHookInvocation(node)) {
+                    const currentHooksIdentifier = hooksStack[hooksStack.length - 1];
+
+                    if (!currentHooksIdentifier || currentHooksIdentifier !== node.callee.object.name) {
+                        context.report({
+                            node: node,
+                            messageId: "noHooksFromAncestorModules",
+                            data: {
+                                context: contextStack[contextStack.length - 1]
+                            }
+                        });
+                    }
+                }
+            },
+
+            "CallExpression:exit": function (node) {
+                if (utils.isModule(node.callee)) {
+                    contextStack.pop();
+                    hooksStack.pop();
```

Comment:
This doesn't look correct. It's possible that a module could use multiple hooks (e.g., beforeEach and afterEach), so having only one `pop` here probably won't always work.

I would suggest having another `if` branch checking to see if the `CallExpression` being exited is a hook, and popping the `hooksStack` at that point.

With the current code, you could write some tests involving both `beforeEach` and `afterEach` hooks in the same module. When you run those, you will probably see some weird behavior with regard to the lint message. You can then use those tests to be more confident that your state management is good. :smile:

Reply:
I think naming is the issue here - I pushed updates where I tried to clarify what I am storing in the stack and when